### DISCUSSION TITLE
pip3 install -e . run errors

### DISCUSSION THREAD

**Discussion by idiomaticrefactoring** [horovod/horovod#3331] [Q&A]
**Checklist:**
yes

**Bug report:**
**Hello I want to run existing test suites, so I run command:**
**pip3 install -e .**
**python3 -m pytest -v test/single/test_run.py**

**However, it occur errors, Actually, I am weird that why I run "pip3 install -e .", it does not install TensorFlow or other packages?**

Collecting cloudpickle
  Using cached cloudpickle-2.0.0-py3-none-any.whl (25 kB)
Collecting psutil
  Using cached psutil-5.8.0-cp37-cp37m-manylinux2010_x86_64.whl (296 kB)
Collecting pyyaml
  Using cached PyYAML-6.0-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (596 kB)
Collecting cffi>=1.4.0
  Using cached cffi-1.15.0-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (427 kB)
Collecting pycparser
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: cloudpickle, psutil, pyyaml, pycparser, cffi, horovod
  Running setup.py develop for horovod
    ERROR: Command errored out with exit status 1:
     command: /mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/mnt/zejun/smp/data/python_star_2000repo/horovod/setup.py'"'"'; __file__='"'"'/mnt/zejun/smp/data/python_star_2000repo/horovod/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: /mnt/zejun/smp/data/python_star_2000repo/horovod/
    Complete output (97 lines):
    running develop
    running egg_info
    creating horovod.egg-info
    writing horovod.egg-info/PKG-INFO
    writing dependency_links to horovod.egg-info/dependency_links.txt
    writing entry points to horovod.egg-info/entry_points.txt
    writing requirements to horovod.egg-info/requires.txt
    writing top-level names to horovod.egg-info/top_level.txt
    writing manifest file 'horovod.egg-info/SOURCES.txt'
    reading manifest file 'horovod.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    warning: no files found matching '*.hpp' under directory '*'
    no previously-included directories found matching '.eggs'
    warning: no directories found matching 'third_party/eigen/Eigen'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/Eigen'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/IterativeLinearSolvers'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/MetisSupport'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/Sparse'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/SparseCholesky'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/SparseLU'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/src/IterativeSolvers/*'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/src/OrderingMethods/Amd.h'
    warning: no previously-included files found matching 'third_party/eigen/Eigen/src/SparseCholesky/*'
    warning: no previously-included files found matching 'third_party/eigen/unsupported/test/mpreal/mpreal.h'
    warning: no previously-included files found matching 'third_party/eigen/unsupported/Eigen/FFT'
    warning: no previously-included files found matching 'third_party/eigen/unsupported/Eigen/MPRealSupport'
    warning: no previously-included files found matching 'third_party/eigen/doc/PreprocessorDirectives.dox'
    warning: no previously-included files found matching 'third_party/eigen/doc/UsingIntelMKL.dox'
    warning: no previously-included files found matching 'third_party/eigen/doc/SparseLinearSystems.dox'
    warning: no previously-included files found matching 'third_party/eigen/COPYING.GPL'
    warning: no previously-included files found matching 'third_party/eigen/COPYING.LGPL'
    warning: no previously-included files found matching 'third_party/eigen/COPYING.README'
    warning: no directories found matching 'third_party/gloo/cmake'
    warning: no files found matching 'CMakeLists.txt' under directory 'third_party/gloo'
    warning: no files found matching '*.in' under directory 'third_party/gloo'
    writing manifest file 'horovod.egg-info/SOURCES.txt'
    running build_ext
    -- Could not find CCache. Consider installing CCache to speed up compilation.
    -- The CXX compiler identification is GNU 7.5.0
    -- Check for working CXX compiler: /usr/bin/c++
    -- Check for working CXX compiler: /usr/bin/c++ -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Build architecture flags: -mf16c -mavx -mfma
    -- Using command /mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7
    -- Could NOT find MPI_CXX (missing: MPI_CXX_LIB_NAMES MPI_CXX_HEADER_DIR MPI_CXX_WORKS)
    -- Could NOT find MPI (missing: MPI_CXX_FOUND)
    -- Could NOT find NVTX (missing: NVTX_INCLUDE_DIR)
    CMake Error at CMakeLists.txt:265 (add_subdirectory):
      The source directory
    
        /mnt/zejun/smp/data/python_star_2000repo/horovod/third_party/gloo
    
      does not contain a CMakeLists.txt file.
    
    
    CMake Error at CMakeLists.txt:267 (target_compile_definitions):
      Cannot specify compile definitions for target "gloo" which is not built by
      this project.
    
    
    -- Could NOT find Tensorflow (missing: Tensorflow_LIBRARIES) (Required is at least version "1.15.0")
    -- Could NOT find Pytorch: Found unsuitable version "", but required is at least "1.2.0" (found )
    -- Could NOT find Mxnet (missing: Mxnet_LIBRARIES) (Required is at least version "1.4.0")
    -- Configuring incomplete, errors occurred!
    See also "/mnt/zejun/smp/data/python_star_2000repo/horovod/build/temp.linux-x86_64-3.7/RelWithDebInfo/CMakeFiles/CMakeOutput.log".
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/mnt/zejun/smp/data/python_star_2000repo/horovod/setup.py", line 213, in <module>
        'horovodrun = horovod.runner.launch:run_commandline'
      File "/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/lib/python3.7/site-packages/setuptools/__init__.py", line 144, in setup
        return distutils.core.setup(**attrs)
      File "/usr/lib/python3.7/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/lib/python3.7/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python3.7/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/lib/python3.7/site-packages/setuptools/command/develop.py", line 38, in run
        self.install_for_development()
      File "/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/lib/python3.7/site-packages/setuptools/command/develop.py", line 140, in install_for_development
        self.run_command('build_ext')
      File "/usr/lib/python3.7/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python3.7/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/lib/python3.7/site-packages/setuptools/command/build_ext.py", line 87, in run
        _build_ext.run(self)
      File "/usr/lib/python3.7/distutils/command/build_ext.py", line 340, in run
        self.build_extensions()
      File "/mnt/zejun/smp/data/python_star_2000repo/horovod/setup.py", line 101, in build_extensions
        cwd=cmake_build_dir)
      File "/usr/lib/python3.7/subprocess.py", line 363, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['cmake', '/mnt/zejun/smp/data/python_star_2000repo/horovod', '-DCMAKE_BUILD_TYPE=RelWithDebInfo', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELWITHDEBINFO=/mnt/zejun/smp/data/python_star_2000repo/horovod/build/lib.linux-x86_64-3.7', '-DPYTHON_EXECUTABLE:FILEPATH=/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7']' returned non-zero exit status 1.
    ----------------------------------------
ERROR: Command errored out with exit status 1: /mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/mnt/zejun/smp/data/python_star_2000repo/horovod/setup.py'"'"'; __file__='"'"'/mnt/zejun/smp/data/python_star_2000repo/horovod/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
WARNING: You are using pip version 20.1.1; however, version 21.3.1 is available.
You should consider upgrading via the '/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7 -m pip install --upgrade pip' command.
Collecting pytest
  Using cached pytest-6.2.5-py3-none-any.whl (280 kB)
Collecting pytest-mock
  Using cached pytest_mock-3.6.1-py3-none-any.whl (12 kB)
Collecting pytest-cov
  Using cached pytest_cov-3.0.0-py3-none-any.whl (20 kB)
Collecting pluggy<2.0,>=0.12
  Using cached pluggy-1.0.0-py2.py3-none-any.whl (13 kB)
Collecting iniconfig
  Using cached iniconfig-1.1.1-py2.py3-none-any.whl (5.0 kB)
Collecting importlib-metadata>=0.12; python_version < "3.8"
  Using cached importlib_metadata-4.10.0-py3-none-any.whl (17 kB)
Collecting toml
  Using cached toml-0.10.2-py2.py3-none-any.whl (16 kB)
Collecting packaging
  Using cached packaging-21.3-py3-none-any.whl (40 kB)
Collecting attrs>=19.2.0
  Using cached attrs-21.2.0-py2.py3-none-any.whl (53 kB)
Collecting py>=1.8.2
  Using cached py-1.11.0-py2.py3-none-any.whl (98 kB)
Collecting coverage[toml]>=5.2.1
  Using cached coverage-6.2-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (213 kB)
Collecting typing-extensions>=3.6.4; python_version < "3.8"
  Using cached typing_extensions-4.0.1-py3-none-any.whl (22 kB)
Collecting zipp>=0.5
  Using cached zipp-3.6.0-py3-none-any.whl (5.3 kB)
Collecting pyparsing!=3.0.5,>=2.0.2
  Using cached pyparsing-3.0.6-py3-none-any.whl (97 kB)
Collecting tomli; extra == "toml"
  Using cached tomli-2.0.0-py3-none-any.whl (12 kB)
Installing collected packages: typing-extensions, zipp, importlib-metadata, pluggy, iniconfig, toml, pyparsing, packaging, attrs, py, pytest, pytest-mock, tomli, coverage, pytest-cov
Successfully installed attrs-21.2.0 coverage-6.2 importlib-metadata-4.10.0 iniconfig-1.1.1 packaging-21.3 pluggy-1.0.0 py-1.11.0 pyparsing-3.0.6 pytest-6.2.5 pytest-cov-3.0.0 pytest-mock-3.6.1 toml-0.10.2 tomli-2.0.0 typing-extensions-4.0.1 zipp-3.6.0
WARNING: You are using pip version 20.1.1; however, version 21.3.1 is available.
You should consider upgrading via the '/mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7 -m pip install --upgrade pip' command.
*************************install*************************
*************************test*************************
============================= test session starts ==============================
platform linux -- Python 3.7.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /mnt/zejun/smp/data/python_star_2000repo/horovod/venv_test_7/bin/python3.7
cachedir: .pytest_cache
rootdir: /mnt/zejun/smp/data/python_star_2000repo/horovod, configfile: setup.cfg
plugins: cov-3.0.0, mock-3.6.1
collecting ... collected 0 items / 1 error

==================================== ERRORS ====================================
___________________ ERROR collecting test/single/test_run.py ___________________
ImportError while importing test module '/mnt/zejun/smp/data/python_star_2000repo/horovod/test/single/test_run.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/single/test_run.py:29: in <module>
    import mock
E   ModuleNotFoundError: No module named 'mock'
=========================== short test summary info ============================
ERROR test/single/test_run.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!

**Comment by maxhgerlach** [comment]
Did you `git clone --recursive`?

Also you miss testing dependencies (mock).

https://horovod.readthedocs.io/en/stable/contributors_include.html
